src/main/java/org/nixos/idea/psi/NixStringLiteralEscaper.kt

```diff
@@ -0,0 +1,142 @@
+package org.nixos.idea.psi
+
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.LiteralTextEscaper
+import com.intellij.psi.PsiLanguageInjectionHost
+import org.intellij.lang.annotations.Language
+import org.nixos.idea.psi.impl.AbstractNixString
+
+class NixStringLiteralEscaper(host: AbstractNixString) : LiteralTextEscaper<PsiLanguageInjectionHost>(host) {
+
+    override fun isOneLine(): Boolean = false
+
+    private var outSourceOffsets: IntArray? = null
+
+    override fun getRelevantTextRange(): TextRange {
+        if (myHost.textLength <= 4) return TextRange.EMPTY_RANGE
+        return TextRange.create(2, myHost.textLength - 2)
+    }
+
+    override fun decode(rangeInsideHost: TextRange, outChars: StringBuilder): Boolean {
+        // TODO issue #81 only indented strings supported for now
+        // single line strings require a new decode function because
+        // it uses different escaping mechanisms
+        if (myHost !is NixIndString) return false
+
+        val subText: String = rangeInsideHost.substring(myHost.text)
+        val array = IntArray(subText.length + 1)
+        val success = unescapeAndDecode(subText, outChars, array)
+        outSourceOffsets = array
+        return success
+    }
+
+    override fun getOffsetInHost(offsetInDecoded: Int, rangeInsideHost: TextRange): Int {
+        val offsets = outSourceOffsets ?: throw IllegalStateException("#decode was not called")
+        val result = if (offsetInDecoded < offsets.size) offsets[offsetInDecoded] else -1
+        return result.coerceIn(0..rangeInsideHost.length) + rangeInsideHost.startOffset
+    }
+
+    companion object {
+        /**
+         * Does not consider interpolations so that
+         * they do appear in the guest language and remain when we end up converting back to Nix.
+         *
+         * @returns the minIndent of the string if successful, or null if unsuccessful.
+         */
+        fun unescapeAndDecode(chars: String, outChars: StringBuilder, sourceOffsets: IntArray?): Boolean {
```

Comment:
note: I am not entirely happy with having two implementations to parse/unescape strings now. Anyway, that is not important right now.